src/Samples/SubModelSeq.List/Program.fs

```diff
@@ -0,0 +1,121 @@
+module Elmish.WPF.Samples.SubModelSeq.List.Program
+
+open Elmish
+open Elmish.WPF
+
+
+type InOutMsg<'a, 'b> =
+  | InMsg of 'a
+  | OutMsg of 'b
+
+
+module List =
+
+  let cons head tail = head :: tail
+
+  let mapAtIndex idx f =
+    List.indexed >> List.map (fun (i, a) -> if i = idx then f a else a)
+
+  let removeAtIndex idx ma =
+    [ ma |> List.take idx
+      ma |> List.skip (idx + 1)
+    ] |> List.concat
+
+
+[<AutoOpen>]
+module Counter =
+
+  type Counter =
+    { Count: int
+      StepSize: int }
+
+  type CounterMsg =
+    | Increment
+    | Decrement
+    | SetStepSize of int
+    | Reset
+
+  module Counter =
+
+    let init =
+      { Count = 0
+        StepSize = 1 }
+    
+    let canReset = (<>) init
+    
+    let update msg m =
+      match msg with
+      | Increment -> { m with Count = m.Count + m.StepSize }
+      | Decrement -> { m with Count = m.Count - m.StepSize }
+      | SetStepSize x -> { m with StepSize = x }
+      | Reset -> init
+
+    let bindings () = [
+      "CounterValue" |> Binding.oneWay(fun (_, _, c) -> c.Count)
+      "Increment" |> Binding.cmd(Increment |> InMsg)
+      "Decrement" |> Binding.cmd(Decrement |> InMsg)
+      "StepSize" |> Binding.twoWay(
+        (fun (_, _, c) -> float c.StepSize),
+        (fun v _ -> v |> int |> SetStepSize |> InMsg))
+      "Reset" |> Binding.cmdIf(
+        Reset |> InMsg,
+        (fun (_, _, c) -> canReset c))
+    ]
+
+
+module App =
+
+  type Model = Counter list
+
+  type Msg =
+    | CounterMsg of int * CounterMsg
+    | AddCounter
+    | Remove of int
+
+  type OutMsg =
+    | OutRemove
+
+
+  let init () =
+    [ Counter.init ]
+
+  let update = function
+    | CounterMsg (idx, msg) -> msg |> Counter.update |> List.mapAtIndex idx
+    | AddCounter -> Counter.init |> List.cons
+    | Remove idx -> idx |> List.removeAtIndex
+
+  let mapOutMsg = function
+    | OutRemove -> Remove
+
+
+module Bindings =
+
+  open App
+
+  let counterBindings () : Binding<Model * int * Counter, InOutMsg<CounterMsg, OutMsg>> list =
+    [ "Remove" |> Binding.cmd(OutRemove |> OutMsg)
+    ] @ (Counter.bindings ())
+
```

Comment:
To stick to what seems like the "canonical" translator pattern, I suggest that `Remove` should be part of `Counter.bindings`, and that these bindings accept the `InMsg` and `OutMsg` functions/constructors as parameters (either as individual parameters, a normal record, or an anonymous record, feel free to experiment).

Reply:
I don't care so much where we put things in the `master` version of this sample.  We when port this to `v4` though, I think we should remove the duplication present in the current branch by removing `InMsg` in all the places it is currently being used and replace `Counter.bindings ()` with
```fsharp
Counter.bindings () |> Bindings.mapMsg InMsg
```

Then case constructors `InMsg` and `OutMsg` are used right next to each other.